### PR TITLE
Use std::chrono for Main Loop Timing

### PR DESCRIPTION
**Pull request by RobertBColton**
I generalized the timing code using high precision `std::chrono` in place of the performance counters. I was originally going to use [SDL_GetPerformanceCounter](https://wiki.libsdl.org/SDL_GetPerformanceCounter) and [SDL_GetPerformanceFrequency](https://wiki.libsdl.org/SDL_GetPerformanceFrequency) to generalize only the timers themselves. However, C++11 has been around long enough that I don't need to and fundies said Android even probably supports chrono at this point.

I have no idea if any of @forthevin's original logic can be further simplified. I do know there are comments left in here directed at @JoshDreamland about Win32 and Xlib's relative order of updating the timer and processing events. I did have to add a 4th timer start variable, due to the nature of chrono being wall-clock based epoch unlike the performance counters, to know how much time was elapsed since the start of the game.

### CODE REVIEW

**Review by JoshDreamland — COMMENTED** [review]
Code-wise, this change *looks* great, but how does it perform? I remember Windows in particular being very finicky with timing. Also, I remember the logic behind one of these systems just working more smoothly than the others, and I'm not sure if you've kept the good logic or the bad. I think this is one we'll need performance metrics on.

**Comment by RobertBColton** [comment]
The one I kept was forthevin's good implementation, which was originally good on Windows, it was xlib that sucked. However, the problem may be related to that comment directed at you @JoshDreamland about the relative order of timing and processing events, which was different between those two platforms in the past. That said, yeah, I guess this could stand to be benchmarked a little, and I'll get back with that data.

I can also add here that Windows actually backs `std::chrono` with the performance counters now anyway. So this code hasn't really changed except one layer of indirection in the C++ runtime.
Windows Chrono Backed by QueryPerformanceCounter

**Comment by RobertBColton** [comment]
I want to note this here that I went looking into the history of `current_time_mcs` and stumbled upon some other facts about regular `current_time` in GM. The first is that it seems to be reflexive in both GM8 and GMSv1.4, always providing monotonic values. And it seems GMSv1.4 broke `current_time` technically because it's no longer milliseconds since the operating system was started, but since the game started, while the manual still states otherwise. In either case, ENIGMA is not now nor was it before respecting either of those properties as far as compatibility is concerned.

```gml
// strictly monotonic reflexive reads
show_message(string(current_time));
show_message(string(current_time));
show_message(string(current_time));
```

**Comment by RobertBColton** [comment]
So I went to do some testing and check things out. I started with Project Mario, testing different frame locks above and below my monitor refresh rate and vsync. Everything seemed fine at first glance, but that was when I noticed some micro-stuttering.

Upon further inspection (a64b2ac626390b67940fdcb8ff5a05ad993a3214), it turned out that I had forgot to clamp the offset deltas. Then I noticed (b06344093b401832416388d81afbf5a984c9725d) that I had also forgot to properly modulus the offset second.

| VSync On                                                                                                                | Microstutter                                                                                                                   | Vsync Off                                                                                                                  |
|-------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------|
| Game Bar Vsync | Game Bar Microstutter | Game Bar No VSync |

**Comment by RobertBColton** [comment]
I did some additional testing with [this old sprite benchmark](https://github.com/enigma-dev/enigma-dev/issues/1617) to take a better look at performance. Frame rates are largely unchanged for SDL and Win32 between this pull request and master. What I did notice comparing further with GM8 and GMSv1.4, is that GM's FPS limiter seems to not be that great. It periodically dips to 59 fps when either capped at 60 fps or vsync is enabled. I noticed some bugs in GMSv1.4 too where I could get way above 60 fps but limiting it to that with vsync was only giving me 45 fps~ for some reason.

**Comment by RobertBColton** [comment]
I modified the 3d cubes example so that you can alter the variable load at runtime.

Download: [3d-cubes-dynamic.zip](https://github.com/enigma-dev/enigma-dev/files/7021340/3d-cubes-dynamic.zip)
Space: Add 100 cubes
Backspace: Remove 100 cubes
V key: toggle vsync

It didn't really seem to behave much different than master. I also did some more testing to make sure `get_timer()` properly behaves for benchmarks. It seems to be consistent as expected and similar to master on the same tests.

@JoshDreamland Considering this, I think we should approve/merge this if there are no other concerns.

**Review by JoshDreamland — APPROVED** [review]
Surprised how clean a change this is. Part of me is considering asking you to keep the other implementations around as backup so that we know where to find them. Maybe just link to this PR in case we need to dig them up one day?

**Pull request merged**
